cmd/ido2icingadb/convert.go

```diff
@@ -0,0 +1,843 @@
+	ObjecttypeId       uint8
+	Name1              string
+	Name2              string
+}
+
+func convertNotificationRows(
+	env string, envId, endpointId icingadbTypes.Binary,
+	selectCache func(dest interface{}, query string, args ...interface{}), ido *sqlx.Tx, idoRows []notificationRow,
+) (icingaDbInserts, _ [][]contracts.Entity, checkpoint any) {
+	if len(idoRows) < 1 {
+		return
+	}
+
+	var cached []struct {
+		HistoryId         uint64
+		PreviousHardState uint8
+	}
+	selectCache(
+		&cached, "SELECT history_id, previous_hard_state FROM previous_hard_state WHERE history_id BETWEEN ? AND ?",
+		idoRows[0].NotificationId, idoRows[len(idoRows)-1].NotificationId,
+	)
+
+	cachedById := make(map[uint64]uint8, len(cached))
+	for _, c := range cached {
+		cachedById[c.HistoryId] = c.PreviousHardState
+	}
+
+	var contacts []struct {
+		NotificationId uint64
+		Name1          string
+	}
+
+	{
+		var query = ido.Rebind(
+			"SELECT c.notification_id, o.name1 FROM icinga_contactnotifications c " +
+				"INNER JOIN icinga_objects o ON o.object_id=c.contact_object_id WHERE c.notification_id BETWEEN ? AND ?",
+		)
+
+		err := ido.Select(&contacts, query, idoRows[0].NotificationId, idoRows[len(idoRows)-1].NotificationId)
+		if err != nil {
+			log.With("query", query).Fatalf("%+v", errors.Wrap(err, "can't perform query"))
+		}
+	}
+
+	contactsById := map[uint64]map[string]struct{}{}
+	for _, contact := range contacts {
+		perId, ok := contactsById[contact.NotificationId]
+		if !ok {
+			perId = map[string]struct{}{}
+			contactsById[contact.NotificationId] = perId
+		}
+
+		perId[contact.Name1] = struct{}{}
+	}
+
+	var notificationHistory, userNotificationHistory, allHistory []contracts.Entity
+	for _, row := range idoRows {
+		checkpoint = row.NotificationId
+
+		previousHardState, ok := cachedById[row.NotificationId]
+		if !ok {
+			continue
+		}
+
+		// The IDO tracks only sent notifications, but not notification config objects, nor even their names.
+		// We have to improvise. By the way we avoid unwanted collisions between synced and migrated data via "ID"
+		// instead of "HOST[!SERVICE]!NOTIFICATION" (ok as this name won't be parsed, but only hashed) and between
+		// migrated data itself via the history ID as object name, i.e. one "virtual object" per sent notification.
+		name := strconv.FormatUint(row.NotificationId, 10)
+
+		nt := convertNotificationType(row.NotificationReason, row.State)
+
+		ntEnum, err := nt.Value()
+		if err != nil {
+			continue
+		}
+
+		ts := convertTime(row.EndTime, row.EndTimeUsec)
+		tsMilli := float64(ts.Time().UnixMilli())
+		notificationHistoryId := hashAny([]interface{}{env, name, ntEnum, tsMilli})
+		id := hashAny([]interface{}{env, "notification", name, ntEnum, tsMilli})
+		typ := objectTypes[row.ObjecttypeId]
+		hostId := calcObjectId(env, row.Name1)
+		serviceId := calcServiceId(env, row.Name1, row.Name2)
+
+		text := row.Output
+		if row.LongOutput.Valid {
+			text += "\n\n" + row.LongOutput.String
+		}
```

Comment:
This can result in an empty string which icingadb-web can't handle at the moment so that it throws the following stack trace at you while scrolling through the notification history:

<details>
<summary>Output not rendered yet</summary>

```
#0 /usr/share/icingaweb2/modules/icingadb/library/Icingadb/Widget/PluginOutputContainer.php(19): Icinga\Module\Icingadb\Util\PluginOutput->isHtml()
#1 [internal function]: Icinga\Module\Icingadb\Widget\PluginOutputContainer->Icinga\Module\Icingadb\Widget\{closure}()
#2 /usr/share/icinga-php/ipl/vendor/ipl/html/src/Attributes.php(415): call_user_func()
#3 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(324): ipl\Html\Attributes->render()
#4 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(499): ipl\Html\BaseHtmlElement->renderUnwrapped()
#5 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(339): ipl\Html\HtmlDocument->render()
#6 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(255): ipl\Html\HtmlDocument->renderUnwrapped()
#7 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(323): ipl\Html\BaseHtmlElement->renderContent()
#8 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(499): ipl\Html\BaseHtmlElement->renderUnwrapped()
#9 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(339): ipl\Html\HtmlDocument->render()
#10 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(255): ipl\Html\HtmlDocument->renderUnwrapped()
#11 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(323): ipl\Html\BaseHtmlElement->renderContent()
#12 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(499): ipl\Html\BaseHtmlElement->renderUnwrapped()
#13 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(339): ipl\Html\HtmlDocument->render()
#14 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(255): ipl\Html\HtmlDocument->renderUnwrapped()
#15 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(323): ipl\Html\BaseHtmlElement->renderContent()
#16 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(499): ipl\Html\BaseHtmlElement->renderUnwrapped()
#17 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(339): ipl\Html\HtmlDocument->render()
#18 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(255): ipl\Html\HtmlDocument->renderUnwrapped()
#19 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(323): ipl\Html\BaseHtmlElement->renderContent()
#20 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(499): ipl\Html\BaseHtmlElement->renderUnwrapped()
#21 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(339): ipl\Html\HtmlDocument->render()
#22 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(255): ipl\Html\HtmlDocument->renderUnwrapped()
#23 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(323): ipl\Html\BaseHtmlElement->renderContent()
#24 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(499): ipl\Html\BaseHtmlElement->renderUnwrapped()
#25 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(339): ipl\Html\HtmlDocument->render()
#26 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(255): ipl\Html\HtmlDocument->renderUnwrapped()
#27 /usr/share/icinga-php/ipl/vendor/ipl/html/src/BaseHtmlElement.php(323): ipl\Html\BaseHtmlElement->renderContent()
#28 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(499): ipl\Html\BaseHtmlElement->renderUnwrapped()
#29 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(339): ipl\Html\HtmlDocument->render()
#30 /usr/share/icinga-php/ipl/vendor/ipl/html/src/HtmlDocument.php(499): ipl\Html\HtmlDocument->renderUnwrapped()
#31 /usr/share/icinga-php/ipl/vendor/ipl/web/src/Compat/ViewRenderer.php(56): ipl\Html\HtmlDocument->render()
#32 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/Helper/ViewRenderer.php(964): ipl\Web\Compat\ViewRenderer->render()
#33 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/HelperBroker.php(272): Zend_Controller_Action_Helper_ViewRenderer->postDispatch()
#34 /usr/share/icingaweb2/modules/icingadb/library/Icingadb/Web/Controller.php(506): Zend_Controller_Action_HelperBroker->notifyPostDispatch()
#35 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Icinga\Module\Icingadb\Web\Controller->dispatch()
#36 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch()
#37 /usr/share/php/Icinga/Application/Web.php(290): Zend_Controller_Front->dispatch()
#38 /usr/share/php/Icinga/Application/webrouter.php(105): Icinga\Application\Web->dispatch()
#39 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#40 {main}
```
</details>

@nilmerg asked to use a dummy text instead. I'd suggest `"-"`.

Reply:
No, please don't. I changed my mind. Web should and can handle this gracefully, it does it already in other cases.

Reply:
https://github.com/Icinga/icingadb-web/pull/666